# PD01 · Explorar y preparar AGROSAVIA

[Índice y distribución de las cuatro horas](../ejercicios/README.md) · [Guía de pandas](../docs/manual-pandas.md)

**Tiempo de trabajo: 60 minutos.** El objetivo es reconocer la estructura de un DataFrame y preparar medidas y fechas conservando los valores originales.

| Símbolo | Uso |
|:---:|---|
| 📘 | Concepto o función de apoyo. |
| ✅ | Evidencia que debe quedar visible. |
| ⚠️ | Límite de interpretación. |

Esta plantilla se encuentra en `ejercicios/PD01_exploracion_agrosavia.ipynb`. Antes de comenzar, se crea una copia en `soluciones/PD01_exploracion_agrosavia.ipynb` dentro del repositorio de cada estudiante. El desarrollo y la entrega se realizan en esa copia; la plantilla de `ejercicios/` se conserva como enunciado. En JupyterLab se puede duplicar el archivo y mover la copia a `soluciones/`, manteniendo el nombre indicado; si ya existe una solución, se continúa sobre ella. Las instrucciones se presentan en celdas Markdown; las celdas de código marcadas como **Desarrollo** contienen el espacio de trabajo. Cada actividad incluye una celda para explicar el resultado.

Se utiliza el kernel del entorno `.venv` del curso. Los CSV ya deben estar descargados mediante `kit/00_datos.py`. La celda inicial prepara las rutas y carga únicamente las entradas de este ejercicio; el desarrollo del análisis corresponde al estudiante. Este notebook se ejecuta de forma independiente.

**Apoyo en el manual:** secciones 1–5 y la receta de registros en revisión de la sección 11.

## Ubicación de trabajo: la raíz del repositorio

📘 **La raíz del repositorio** es la carpeta principal de la copia del proyecto. En la instalación del curso se llama `big-data-postgraduate` y contiene `README.md`, `kit/`, `ejercicios/` y `soluciones/`:

```text
bigdata/
└── big-data-postgraduate/       ← raíz del repositorio
    ├── README.md
    ├── kit/
    ├── ejercicios/              ← enunciados y plantillas
    └── soluciones/              ← notebooks resueltos
```

**Desde la terminal de Ubuntu en WSL.** `pwd` muestra la carpeta actual, `cd` cambia de carpeta y `ls` muestra su contenido. Los siguientes comandos se ejecutan en la terminal, no en una celda Python:

```bash
pwd
cd /mnt/c/Users/TUPTC/bigdata/big-data-postgraduate
pwd
ls
```

✅ Después de `cd`, el segundo `pwd` debe mostrar `/mnt/c/Users/TUPTC/bigdata/big-data-postgraduate`. En la salida de `ls` deben aparecer `README.md`, `kit`, `ejercicios` y `soluciones`. La carpeta `bigdata` es la carpeta contenedora; la raíz del proyecto es su subcarpeta `big-data-postgraduate`.

La ruta anterior corresponde a la configuración de la guía WSL. Si el repositorio se clonó en otra ubicación, se sustituye por la ruta real de esa copia. Si `cd` informa que la carpeta no existe, se revisa esa ubicación antes de continuar.

**Desde una subcarpeta conocida.** `..` significa «carpeta superior». Se comprueba primero la ubicación con `pwd` y se utiliza la alternativa correspondiente:

| Ubicación actual, relativa al repositorio | Comando para volver a la raíz |
|---|---|
| `ejercicios/`, `soluciones/` o `kit/` | `cd ..` |
| `kit/Notebooks/` | `cd ../..` |

Al llegar a la raíz, se puede activar el entorno y abrir JupyterLab desde la terminal:

```bash
source .venv/bin/activate
jupyter lab
```

**Dentro de un notebook ya abierto.** La terminal y el kernel de Python mantienen sus propios directorios de trabajo. La celda de carga siguiente localiza la raíz y la guarda en `REPO`, aunque el notebook esté abierto desde `soluciones/` o `ejercicios/`. Después de ejecutar esa celda, el siguiente bloque puede ejecutarse en una celda Python para comprobar la ubicación y cambiar el directorio del kernel a la raíz:

```python
import os
from pathlib import Path

print("Directorio antes del cambio:", Path.cwd())
print("Raíz detectada:", REPO)
os.chdir(REPO)
print("Directorio después del cambio:", Path.cwd())
```

✅ `Path.cwd()` debe coincidir con `REPO` después del cambio. Esta operación cambia la base para interpretar rutas relativas; el archivo del notebook permanece en `soluciones/`. La carga de datos de estos ejercicios también funciona sin ese cambio porque construye las rutas a partir de `REPO`.

## Preparación

Se ejecuta la siguiente celda y se comprueba que el intérprete pertenece al entorno del curso y que pandas corresponde a la versión 2.2.3. Las rutas se calculan desde el repositorio, sin introducir rutas personales de Windows, WSL o macOS.

In [2]:
from pathlib import Path
import sys
import pandas as pd
from IPython.display import display

actual = Path.cwd().resolve()
REPO = next(
    (ruta for ruta in [actual, *actual.parents]
     if (ruta / "kit/fuentes.json").is_file()
     and (ruta / "kit/00_datos.py").is_file()),
    None,
)
if REPO is None:
    raise FileNotFoundError("El notebook debe abrirse dentro del repositorio del curso")

RAW = REPO / "kit/data/raw"
OUT = REPO / "kit/salidas/soluciones/PD01"
archivos = ['agrosavia.csv']
faltantes = [nombre for nombre in archivos if not (RAW / nombre).is_file()]
if faltantes:
    raise FileNotFoundError(f"Faltan CSV en {RAW}: {faltantes}")
OUT.mkdir(parents=True, exist_ok=True)

opciones = dict(sep=",", encoding="utf-8-sig", dtype=str, keep_default_na=False)
suelo = pd.read_csv(RAW / "agrosavia.csv", **opciones)


print("Python:", sys.version.split()[0])
print("Intérprete:", sys.executable)
print("pandas:", pd.__version__)
print("Repositorio:", REPO)
print("Salidas:", OUT)

Python: 3.12.15
Intérprete: /mnt/c/Users/TUPTC/bigdata/big-data-postgraduate/.venv/bin/python
pandas: 2.2.3
Repositorio: /mnt/c/Users/TUPTC/bigdata/big-data-postgraduate
Salidas: /mnt/c/Users/TUPTC/bigdata/big-data-postgraduate/kit/salidas/soluciones/PD01


## 1. Inspección inicial · 10 minutos

Presentar la cantidad de filas y columnas, los nombres de las columnas, sus tipos y cinco registros. Explicar qué representa una fila y por qué la lectura inicial conserva los campos como texto.

📘 **Apoyo:** `shape`, `columns`, `dtypes`, `head` e `info`.

✅ **Comprobación:** El corte del manifiesto contiene 92.738 filas y 32 columnas. La tabla original debe conservar esa cantidad de filas durante el ejercicio.

In [4]:
# Desarrollo de la actividad 1
total_original = len(suelo)

print("Dimensiones (filas, columnas):", suelo.shape)
print("¿Coincide con el manifiesto (92.738 x 32)?", suelo.shape == (92738, 32))
print("\nColumnas:")
for i, nombre in enumerate(suelo.columns, start=1):
    print(f"{i:>2}. {nombre}")

print("\nTipos de dato por cantidad de columnas:")
print(suelo.dtypes.astype(str).value_counts())
print()

display(suelo.head(5))
suelo.info(memory_usage="deep")

Dimensiones (filas, columnas): (92738, 32)
¿Coincide con el manifiesto (92.738 x 32)? True

Columnas:
 1. Secuencial
 2. Fecha de Análisis
 3. Departamento
 4. Municipio
 5. Cultivo
 6. Estado
 7. Tiempo de establecimiento
 8. Topografia
 9. Drenaje
10. Riego
11. Fertilizantes aplicados
12. pH agua:suelo
13. Materia organica
14. Fósforo Bray II
15. Azufre Fosfato monocalcico
16. Acidez Intercambiable
17. Aluminio intercambiable
18. Calcio intercambiable
19. Magnesio intercambiable
20. Potasio intercambiable
21. Sodio intercambiable
22. capacidad de intercambio cationico
23. Conductividad electrica
24. Hierro disponible olsen
25. Cobre disponible
26. Manganeso disponible Olsen
27. Zinc disponible Olsen
28. Boro disponible
29. Hierro disponible doble acido
30. Cobre disponible doble acido
31. Manganeso disponible doble acido
32. Zinc disponible doble  acido

Tipos de dato por cantidad de columnas:
object    32
Name: count, dtype: int64



,Secuencial,Fecha de Análisis,Departamento,Municipio,Cultivo,Estado,Tiempo de establecimiento,Topografia,Drenaje,Riego,...,Conductividad electrica,Hierro disponible olsen,Cobre disponible,Manganeso disponible Olsen,Zinc disponible Olsen,Boro disponible,Hierro disponible doble acido,Cobre disponible doble acido,Manganeso disponible doble acido,Zinc disponible doble acido
0,1,7/01/2014,NARIÑO,SAN ANDRÉS DE TUMACO,No Indica,No indica,No indica,No indica,No indica,No indica,...,0.133,66.39,1.700,1.700,1.5,0.187,ND,ND,ND,ND
1,2,21/08/2014,HUILA,SANTA MARÍA,Granadilla,Establecido,De 1 a 5 años,Pendiente,Buen drenaje,No Tiene,...,2.749,250.9,3.7,29.4,28.70,2.084,ND,ND,ND,ND
2,3,22/08/2014,ANTIOQUIA,LIBORINA,Café,Por establecer,No indica,Pendiente,Mal drenaje,No Tiene,...,0.328,390,11.10,8.299,5,0.085,ND,ND,ND,ND
3,4,22/08/2014,ANTIOQUIA,LIBORINA,Maracuyá,Por establecer,No indica,Ondulado,Mal drenaje,No Tiene,...,0.171,200,4.2,4.699,1,0.229,ND,ND,ND,ND
4,5,22/08/2014,ANTIOQUIA,LIBORINA,Café,Por establecer,No indica,Ondulado,Mal drenaje,No Tiene,...,0.323,117.9,3.8,4.600,0.700,0.139,ND,ND,ND,ND


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 92738 entries, 0 to 92737
Data columns (total 32 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   Secuencial                          92738 non-null  object
 1   Fecha de Análisis                   92738 non-null  object
 2   Departamento                        92738 non-null  object
 3   Municipio                           92738 non-null  object
 4   Cultivo                             92738 non-null  object
 5   Estado                              92738 non-null  object
 6   Tiempo de establecimiento           92738 non-null  object
 7   Topografia                          92738 non-null  object
 8   Drenaje                             92738 non-null  object
 9   Riego                               92738 non-null  object
 10  Fertilizantes aplicados             92738 non-null  object
 11  pH agua:suelo                       92738 non-null  ob

**Interpretación del resultado:**

La tabla tiene **92.738 filas y 32 columnas**, igual que el corte del manifiesto. Cada fila representa **un análisis de suelo de laboratorio**, identificado por `Secuencial`, con su fecha, el departamento y municipio de procedencia y los resultados de las variables medidas (pH, materia orgánica, nutrientes, etc.). Una fila no es un predio ni un punto georreferenciado.

Todas las columnas aparecen como `object` (texto) porque se leyeron con `dtype=str` y `keep_default_na=False`. Es intencional: así se conservan los ceros iniciales de los códigos, la coma decimal de valores como `5,6` y los marcadores como `ND`, los vacíos o los textos del tipo `>10.00`, que pandas habría convertido automáticamente. Primero se diagnostica qué contiene cada campo y después se convierten solo las columnas que de verdad son numéricas o fechas. El total original queda guardado en `total_original` para los controles posteriores.

## 2. Diagnóstico de cinco campos · 10 minutos

Construir una tabla llamada `perfil` para `Secuencial`, `Fecha de Análisis`, `Departamento`, `Municipio` y `pH agua:suelo`. Para cada campo se registran los nulos de pandas, los textos vacíos después de retirar espacios y los marcadores `ND`. Las tres situaciones se cuentan por separado.

📘 **Apoyo:** `isna`, `.str.strip`, `.str.upper`, `.eq` y `sum`.

✅ **Comprobación:** `perfil` debe tener una fila por campo. Con `keep_default_na=False`, un texto vacío o `ND` no se convierte automáticamente en un nulo.

In [5]:
# Desarrollo de la actividad 2
campos = ["Secuencial", "Fecha de Análisis", "Departamento", "Municipio", "pH agua:suelo"]

filas = []
for campo in campos:
    columna = suelo[campo]
    limpio = columna.str.strip()
    filas.append({
        "campo": campo,
        "nulos_pandas": int(columna.isna().sum()),
        "vacios_texto": int(limpio.eq("").sum()),
        "marcador_ND": int(limpio.str.upper().eq("ND").sum()),
        "distintos": int(columna.nunique(dropna=False)),
    })

perfil = pd.DataFrame(filas).set_index("campo")
assert len(perfil) == len(campos)
display(perfil)

,nulos_pandas,vacios_texto,marcador_ND,distintos
campo,,,,
Secuencial,0,0,0,92738
Fecha de Análisis,0,0,0,2388
Departamento,0,0,0,33
Municipio,0,2,0,991
pH agua:suelo,0,0,0,737


**Interpretación del resultado:**

`perfil` tiene una fila por cada uno de los cinco campos y separa tres situaciones distintas: nulos de pandas, textos vacíos y marcadores `ND`.

- **Nulos de pandas:** 0 en los cinco campos. Como se leyó con `keep_default_na=False`, pandas no crea `NaN` al leer: las ausencias quedan escondidas como texto. Si solo se usara `isna()`, el perfil diría que el archivo está completo.
- **Textos vacíos:** solo `Municipio` tiene **2 registros vacíos** (`""`). Esos 2 análisis no tienen municipio de procedencia y deberían revisarse antes de cualquier agregación municipal. Por eso `Municipio` muestra 991 valores distintos: 990 municipios más el texto vacío.
- **Marcadores `ND`:** 0 en los cinco campos.

Los demás campos están completos. `Secuencial` tiene 92.738 valores distintos, uno por fila, así que identifica de forma única cada análisis. `Fecha de Análisis` tiene 2.388 fechas distintas y `Departamento` 33 valores, sin ausencias. `pH agua:suelo` no tiene vacíos ni `ND` (737 valores distintos).

## 3. Conversión de pH y fecha · 15 minutos

Crear `ph` desde `pH agua:suelo` y `fecha_dt` desde `Fecha de Análisis`. Utilizar conversión numérica con `errors="coerce"` y fecha con formato `%d/%m/%Y`. Conservar los originales. Distinguir entradas vacías de textos no vacíos que no pudieron convertirse.

📘 **Apoyo:** `pd.to_numeric`, `.str.replace`, `pd.to_datetime`, `isna` y máscaras booleanas.

✅ **Comprobación:** Se muestran los conteos de conversiones fallidas y hasta cinco ejemplos con su valor original. No se reemplazan medidas ausentes por cero.

In [6]:
# Desarrollo de la actividad 3
ph_texto = suelo["pH agua:suelo"].str.strip()
fecha_texto = suelo["Fecha de Análisis"].str.strip()

# Columnas nuevas; las originales no se modifican.
suelo["ph"] = pd.to_numeric(ph_texto.str.replace(",", ".", regex=False), errors="coerce")
suelo["fecha_dt"] = pd.to_datetime(fecha_texto, format="%d/%m/%Y", errors="coerce")

ph_vacio = ph_texto.eq("")
ph_no_numerico = ~ph_vacio & suelo["ph"].isna()
fecha_vacia = fecha_texto.eq("")
fecha_no_valida = ~fecha_vacia & suelo["fecha_dt"].isna()

fallos_conversion = pd.DataFrame({
    "entrada_vacia": [int(ph_vacio.sum()), int(fecha_vacia.sum())],
    "texto_no_convertible": [int(ph_no_numerico.sum()), int(fecha_no_valida.sum())],
}, index=["ph", "fecha_dt"])
fallos_conversion["total_ausente"] = [int(suelo["ph"].isna().sum()), int(suelo["fecha_dt"].isna().sum())]
display(fallos_conversion)

print("Ejemplos de pH no convertible (valor original):")
display(suelo.loc[ph_no_numerico, ["Secuencial", "pH agua:suelo"]].head(5))
print("Ejemplos de fecha no interpretable (valor original):")
display(suelo.loc[fecha_no_valida, ["Secuencial", "Fecha de Análisis"]].head(5))

print("Rango de pH convertido:", suelo["ph"].min(), "a", suelo["ph"].max())
print("Rango de fechas:", suelo["fecha_dt"].min(), "a", suelo["fecha_dt"].max())
assert len(suelo) == total_original

,entrada_vacia,texto_no_convertible,total_ausente
ph,0,11,11
fecha_dt,0,0,0


Ejemplos de pH no convertible (valor original):


,Secuencial,pH agua:suelo
7282,7283,4..01
46528,46516,>10.00
46530,46518,>10.00
52903,52901,<3.80
56330,56329,<3.80


Ejemplos de fecha no interpretable (valor original):


,Secuencial,Fecha de Análisis


Rango de pH convertido: 3.19 a 10.57
Rango de fechas: 2014-01-07 00:00:00 a 2025-09-05 00:00:00


**Interpretación del resultado:**

Se crearon `ph` (numérico, cambiando la coma decimal por punto) y `fecha_dt` (fecha con formato explícito `%d/%m/%Y`), sin modificar las columnas originales `pH agua:suelo` y `Fecha de Análisis`.

- **pH:** 0 entradas vacías y **11 textos no convertibles**. No faltan datos: el valor existe pero no es un número. Los ejemplos muestran dos tipos de caso:
  - **Valores censurados**, como `>10.00` y `<3.80`: el laboratorio informa que el pH está por encima o por debajo del límite que mide el método, pero no da el valor exacto.
  - **Errores de digitación**, como `4..01` (doble punto), que probablemente correspondía a 4,01.
- **Fecha:** 0 vacías y 0 no interpretables. Todas las fechas siguen el formato día/mes/año.



## 4. Registros para revisión · 15 minutos

Crear la columna booleana `en_revision`. Se marca una fila si su pH es ausente, está fuera de `[0, 14]` o su fecha no se pudo interpretar. Contar por separado cada motivo y el total de filas marcadas. Conservar todas las observaciones.

📘 **Apoyo:** `between`, `isna`, `|`, `~`, `.loc` y `value_counts`.

✅ **Comprobación:** Las filas marcadas y no marcadas deben sumar el total original. Una misma fila puede tener varios motivos; por eso la suma de los motivos no necesariamente coincide con el total marcado.

In [7]:
# Desarrollo de la actividad 4
motivo_ph_ausente = suelo["ph"].isna()
motivo_ph_fuera = suelo["ph"].notna() & ~suelo["ph"].between(0, 14)
motivo_fecha = suelo["fecha_dt"].isna()

suelo["en_revision"] = motivo_ph_ausente | motivo_ph_fuera | motivo_fecha

motivos = pd.Series({
    "pH ausente o no numérico": int(motivo_ph_ausente.sum()),
    "pH fuera de [0, 14]": int(motivo_ph_fuera.sum()),
    "fecha no interpretable": int(motivo_fecha.sum()),
}, name="filas")
display(motivos.to_frame())

conteo_revision = suelo["en_revision"].value_counts().rename("filas")
display(conteo_revision.to_frame())

marcadas = int(suelo["en_revision"].sum())
no_marcadas = int((~suelo["en_revision"]).sum())
print(f"Marcadas: {marcadas:,} | No marcadas: {no_marcadas:,} | Total: {marcadas + no_marcadas:,}")
print("Filas con más de un motivo:",
      int((motivo_ph_ausente.astype(int) + motivo_ph_fuera.astype(int) + motivo_fecha.astype(int)).gt(1).sum()))

assert marcadas + no_marcadas == total_original
assert len(suelo) == total_original
display(suelo.loc[suelo["en_revision"], ["Secuencial", "pH agua:suelo", "ph", "Fecha de Análisis", "fecha_dt"]].head(10))

,filas
pH ausente o no numérico,11
"pH fuera de [0, 14]",0
fecha no interpretable,0


,filas
en_revision,
False,92727
True,11


Marcadas: 11 | No marcadas: 92,727 | Total: 92,738
Filas con más de un motivo: 0


,Secuencial,pH agua:suelo,ph,Fecha de Análisis,fecha_dt
7282,7283,4..01,NaN,8/01/2014,2014-01-08
46528,46516,>10.00,NaN,1/11/2018,2018-11-01
46530,46518,>10.00,NaN,1/11/2018,2018-11-01
52903,52901,<3.80,NaN,7/11/2019,2019-11-07
56330,56329,<3.80,NaN,15/07/2019,2019-07-15
57555,57554,<3.80,NaN,27/08/2020,2020-08-27
61554,61554,>10.00,NaN,4/10/2021,2021-10-04
61556,61556,>10.00,NaN,4/10/2021,2021-10-04
70064,70054,<3.80,NaN,19/08/2022,2022-08-19
72731,72732,<3.80,NaN,2/11/2022,2022-11-02


**Interpretación del resultado:**

`en_revision` marca **11 filas** y deja **92.727 sin marca**. Su suma es 92.738, igual al total original, y el `assert` confirma que **no se eliminó ninguna observación**: las filas quedan señaladas para revisión, no borradas.

Por motivo: **11 por pH ausente o no numérico**, 0 por pH fuera de `[0, 14]` y 0 por fecha no interpretable. Hay **0 filas con más de un motivo**, así que en este caso la suma de motivos coincide con el total marcado; podría no coincidir si una misma fila tuviera, por ejemplo, pH y fecha inválidos a la vez.

La tabla de filas marcadas muestra que todas tienen fecha válida (entre 2014 y 2022) y que el problema está solo en el pH: valores censurados (`>10.00`, `<3.80`) y un error de digitación (`4..01`).

## 5. Evidencias y conclusiones · 10 minutos

Construir `controles` con las columnas `control` y `cantidad`: total de entrada, pH sin número, pH fuera del intervalo, fecha no interpretable y total de filas en revisión. Exportar a `OUT / "controles_agrosavia.csv"` sin índice. Redactar dos hallazgos apoyados en los conteos.

📘 **Apoyo:** `pd.DataFrame`, `to_csv(index=False)` y las tablas de las actividades anteriores.

✅ **Comprobación:** El archivo se genera dentro de la ruta `OUT`. Cada hallazgo menciona un conteo y explica su significado o limitación.

In [8]:
# Desarrollo de la actividad 5
controles = pd.DataFrame({
    "control": [
        "total_entrada",
        "ph_sin_numero",
        "ph_fuera_intervalo",
        "fecha_no_interpretable",
        "total_en_revision",
    ],
    "cantidad": [
        total_original,
        int(motivo_ph_ausente.sum()),
        int(motivo_ph_fuera.sum()),
        int(motivo_fecha.sum()),
        int(suelo["en_revision"].sum()),
    ],
})
display(controles)

ruta = OUT / "controles_agrosavia.csv"
controles.to_csv(ruta, index=False, encoding="utf-8-sig")
assert ruta.is_file()
print("Archivo generado:", ruta)

pct = controles.loc[4, "cantidad"] / total_original * 100
print(f"Filas en revisión: {controles.loc[4, 'cantidad']:,} de {total_original:,} ({pct:.3f} %)")

,control,cantidad
0,total_entrada,92738
1,ph_sin_numero,11
2,ph_fuera_intervalo,0
3,fecha_no_interpretable,0
4,total_en_revision,11


Archivo generado: /mnt/c/Users/TUPTC/bigdata/big-data-postgraduate/kit/salidas/soluciones/PD01/controles_agrosavia.csv
Filas en revisión: 11 de 92,738 (0.012 %)


**Interpretación del resultado:**

`controles` resume la conciliación del ejercicio (92.738 de entrada; 11 sin pH numérico; 0 fuera de intervalo; 0 fechas no interpretables; 11 en revisión) y se exportó a `kit/salidas/soluciones/PD01/controles_agrosavia.csv` sin índice.

**Hallazgo 1.** De 92.738 análisis, solo **11 (≈0,012 %)** no tienen un pH numérico utilizable, y ninguno está fuera del intervalo físico `[0, 14]`. La calidad del campo pH es muy alta: excluir esas filas casi no cambia el tamaño de la muestra. Sin embargo, la mayoría son **valores censurados** (`>10.00`, `<3.80`) que indican suelos extremadamente alcalinos o ácidos. Si se descartan sin más, se pierden justo los casos extremos. Que un valor pase el control `[0, 14]` no prueba que sea correcto agronómicamente; solo que es físicamente posible.

**Hallazgo 2.** Hay **0 fechas no interpretables**, y en los campos de identificación solo aparecen **2 municipios vacíos**; los problemas se concentran en pocos registros. Leer con `dtype=str` y `keep_default_na=False` fue necesario para verlos: con la lectura por defecto, los municipios vacíos se habrían convertido en `NaN` sin distinguir su origen, y los textos como `>10.00` se habrían mezclado con el resto de ausencias. Además, la cantidad de análisis por municipio refleja **dónde se tomaron muestras**, no la distribución real de los suelos del país.

## Entrega

El notebook completado se conserva como `soluciones/PD01_exploracion_agrosavia.ipynb`. La salida regenerable es `kit/salidas/soluciones/PD01/controles_agrosavia.csv`.

✅ Deben quedar visibles el perfil, los controles de conversión, la conciliación de filas y dos hallazgos. El notebook debe ejecutar todas las celdas en orden después de reiniciar el kernel.

⚠️ El intervalo de pH es un control de dominio para esta práctica. No determina por sí solo una recomendación agronómica. La frecuencia de análisis de suelo no representa automáticamente la distribución de los suelos de todo el territorio.